src/librustc_typeck/check/cast.rs

```diff
@@ -457,14 +478,20 @@ impl<'a, 'gcx, 'tcx> CastCheck<'tcx> {
         debug!("check_ptr_ptr_cast m_expr={:?} m_cast={:?}", m_expr, m_cast);
         // ptr-ptr cast. vtables must match.
 
+        let (expr_kind, cast_kind) =
```

Comment:
This forbids casting an `*const _` to an `*const u32`, please move the check after the if.